Mute migrate signals when calling apply_*_migration

maikhoepfel opened the pull request:
This moves signal handling into the Migrator class. As a desired side effect, Migrator.reset is called with unmuted signals.

Fixes https://github.com/wemake-services/django-test-migrations/issues/128

skarzi commented on django_test_migrations/migrator.py:
```diff
@@ -65,12 +66,18 @@ def apply_tested_migration(self, targets: MigrationSpec) -> ProjectState:
 
     def reset(self) -> None:
-        """Reset the state to the most recent one."""
+        """
+        Reset the state to the most recent one.
+
+        Notably, signals are not muted here to avoid
+        https://github.com/wemake-services/django-test-migrations/issues/128
+        """
```
```suggestion
        """ Reset the state to the most recent one.

        Notably, signals are not muted here to avoid
        https://github.com/wemake-services/django-test-migrations/issues/128
        
        """
```

sobolevn replied:
We actually use 

```
"""
Reset the state to the most recent one.
```

style 🙂

skarzi replied:
sorry, my bad - I've corrected the suggestion's code

skarzi commented on django_test_migrations/signals.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@contextmanager
+def mute_migrate_signals():
```
Let's add docstring to avoid flake8 `D103` violation

maikhoepfel replied:
I wanted to, but couldn't figure out what the reason for muting the signals is. My test suite works fine without it, and I couldn't find a discussion of it. 
I'll add a stub and you are welcome to expand it.

skarzi replied:
The only thing that comes to my mind is that some signals receivers/handlers could have some side effect that can potentially cause issues/crashes, so we are disabling it when preparing the testing database for the migration test

skarzi commented on django_test_migrations/signals.py:
```diff
@@ -0,0 +1,13 @@
+def mute_migrate_signals():
+    pre_migrate_receivers = pre_migrate.receivers
+    post_migrate_receivers = post_migrate.receivers
+    pre_migrate.receivers = post_migrate.receivers = []
```
Following assignment will store the same `list` (because `list` is mutable) for both attributes which can lead to many hard do debug errors.

```suggestion
    pre_migrate.receivers = []
    post_migrate.receivers = []
```

maikhoepfel replied:
Great catch, thanks!

skarzi reviewed (commented):
Thanks for the PR!

I just added a few comments with suggestions.

maikhoepfel commented:
Thanks for the kind review, everyone. I've updated the PR with the great suggestions, and also updated the tests to reflect the changes in signal handling.

skarzi reviewed (approved):
thanks, great job :+1:
